File: services/timeline_fonts.py
"""Timeline font catalog — paths mirror ui/frontend/src/lib/timelineFonts.ts."""

from __future__ import annotations

from pathlib import Path

_FONTS_ROOT = (
    Path(__file__).resolve().parents[1] / "ui" / "frontend" / "public" / "fonts"
).resolve()

_CATALOG: dict[str, dict[int, str]] = {
    "inter": {400: "inter/Inter-Regular.ttf", 700: "inter/Inter-Bold.ttf"},
    "roboto": {400: "roboto/Roboto-Regular.ttf", 700: "roboto/Roboto-Bold.ttf"},
    "open-sans": {400: "open-sans/OpenSans-Regular.ttf", 700: "open-sans/OpenSans-Bold.ttf"},
    "lato": {400: "lato/Lato-Regular.ttf", 700: "lato/Lato-Bold.ttf"},
    "montserrat": {400: "montserrat/Montserrat-Regular.ttf", 700: "montserrat/Montserrat-Bold.ttf"},
    "raleway": {400: "raleway/Raleway-Regular.ttf", 700: "raleway/Raleway-Bold.ttf"},
    "poppins": {400: "poppins/Poppins-Regular.ttf", 700: "poppins/Poppins-Bold.ttf"},
    "nunito": {400: "nunito/Nunito-Regular.ttf", 700: "nunito/Nunito-Bold.ttf"},
    "source-sans-3": {400: "source-sans-3/SourceSans3-Regular.ttf", 700: "source-sans-3/SourceSans3-Bold.ttf"},
    "ubuntu": {400: "ubuntu/Ubuntu-Regular.ttf", 700: "ubuntu/Ubuntu-Bold.ttf"},
    "rubik": {400: "rubik/Rubik-Regular.ttf", 700: "rubik/Rubik-Bold.ttf"},
    "oswald": {400: "oswald/Oswald-Regular.ttf", 700: "oswald/Oswald-Bold.ttf"},
    "bebas-neue": {400: "bebas-neue/BebasNeue-Regular.ttf"},
    "playfair-display": {400: "playfair-display/PlayfairDisplay-Regular.ttf", 700: "playfair-display/PlayfairDisplay-Bold.ttf"},
    "merriweather": {400: "merriweather/Merriweather-Regular.ttf", 700: "merriweather/Merriweather-Bold.ttf"},
    "noto-sans": {400: "noto-sans/NotoSans-Regular.ttf", 700: "noto-sans/NotoSans-Bold.ttf"},
    "noto-sans-jp": {400: "noto-sans-jp/NotoSansJP-Regular.ttf", 700: "noto-sans-jp/NotoSansJP-Bold.ttf"},
    "m-plus-rounded-1c": {400: "m-plus-rounded-1c/MPLUSRounded1c-Regular.ttf", 700: "m-plus-rounded-1c/MPLUSRounded1c-Bold.ttf"},
    "pacifico": {400: "pacifico/Pacifico-Regular.ttf"},
    "jetbrains-mono": {400: "jetbrains-mono/JetBrainsMono-Regular.ttf", 700: "jetbrains-mono/JetBrainsMono-Bold.ttf"},
}


def resolve_timeline_font_path(family_id: str, weight: int) -> Path:
    """Return absolute path to a bundled TTF; falls back to Inter Regular."""
    fam = _CATALOG.get(family_id) or _CATALOG["inter"]
    rel = fam.get(weight) or fam.get(400) or next(iter(fam.values()))
    path = (_FONTS_ROOT / rel).resolve()
    if path.is_file():
        return path
    # Any ttf in family folder
    fam_dir = _FONTS_ROOT / family_id
    if fam_dir.is_dir():
        for f in sorted(fam_dir.glob("*.ttf")):
            return f.resolve()
    inter = _FONTS_ROOT / "inter"
    if inter.is_dir():
        for f in sorted(inter.glob("*.ttf")):
            return f.resolve()
    raise FileNotFoundError(f"Timeline font not found: {family_id} weight={weight}")

File: ui/frontend/src/lib/timelineFonts.ts
export type TimelineFontVariant = {
  weight: number;
  style: "normal";
  woff2: string;
  ttf: string;
};

export type TimelineFontFamily = {
  id: string;
  label: string;
  category: string;
  variants: TimelineFontVariant[];
};

function v(weight: number, file: string): TimelineFontVariant {
  const ttf = `/fonts/${file}.ttf`;
  return {
    weight,
    style: "normal",
    woff2: ttf,
    ttf,
  };
}

/** Curated timeline font catalog (~20 families, Regular + Bold). */
export const TIMELINE_FONTS: TimelineFontFamily[] = [
  { id: "inter", label: "Inter", category: "Sans", variants: [v(400, "inter/Inter-Regular"), v(700, "inter/Inter-Bold")] },
  { id: "roboto", label: "Roboto", category: "Sans", variants: [v(400, "roboto/Roboto-Regular"), v(700, "roboto/Roboto-Bold")] },
  { id: "open-sans", label: "Open Sans", category: "Sans", variants: [v(400, "open-sans/OpenSans-Regular"), v(700, "open-sans/OpenSans-Bold")] },
  { id: "lato", label: "Lato", category: "Sans", variants: [v(400, "lato/Lato-Regular"), v(700, "lato/Lato-Bold")] },
  { id: "montserrat", label: "Montserrat", category: "Sans", variants: [v(400, "montserrat/Montserrat-Regular"), v(700, "montserrat/Montserrat-Bold")] },
  { id: "raleway", label: "Raleway", category: "Sans", variants: [v(400, "raleway/Raleway-Regular"), v(700, "raleway/Raleway-Bold")] },
  { id: "poppins", label: "Poppins", category: "Sans", variants: [v(400, "poppins/Poppins-Regular"), v(700, "poppins/Poppins-Bold")] },
  { id: "nunito", label: "Nunito", category: "Sans", variants: [v(400, "nunito/Nunito-Regular"), v(700, "nunito/Nunito-Bold")] },
  { id: "source-sans-3", label: "Source Sans 3", category: "Sans", variants: [v(400, "source-sans-3/SourceSans3-Regular"), v(700, "source-sans-3/SourceSans3-Bold")] },
  { id: "ubuntu", label: "Ubuntu", category: "Sans", variants: [v(400, "ubuntu/Ubuntu-Regular"), v(700, "ubuntu/Ubuntu-Bold")] },
  { id: "rubik", label: "Rubik", category: "Sans", variants: [v(400, "rubik/Rubik-Regular"), v(700, "rubik/Rubik-Bold")] },
  { id: "oswald", label: "Oswald", category: "Display", variants: [v(400, "oswald/Oswald-Regular"), v(700, "oswald/Oswald-Bold")] },
  { id: "bebas-neue", label: "Bebas Neue", category: "Display", variants: [v(400, "bebas-neue/BebasNeue-Regular")] },
  { id: "playfair-display", label: "Playfair Display", category: "Serif", variants: [v(400, "playfair-display/PlayfairDisplay-Regular"), v(700, "playfair-display/PlayfairDisplay-Bold")] },
  { id: "merriweather", label: "Merriweather", category: "Serif", variants: [v(400, "merriweather/Merriweather-Regular"), v(700, "merriweather/Merriweather-Bold")] },
  { id: "noto-sans", label: "Noto Sans", category: "Sans", variants: [v(400, "noto-sans/NotoSans-Regular"), v(700, "noto-sans/NotoSans-Bold")] },
  { id: "noto-sans-jp", label: "Noto Sans JP", category: "Japanese", variants: [v(400, "noto-sans-jp/NotoSansJP-Regular"), v(700, "noto-sans-jp/NotoSansJP-Bold")] },
  { id: "m-plus-rounded-1c", label: "M PLUS Rounded 1c", category: "Japanese", variants: [v(400, "m-plus-rounded-1c/MPLUSRounded1c-Regular"), v(700, "m-plus-rounded-1c/MPLUSRounded1c-Bold")] },
  { id: "pacifico", label: "Pacifico", category: "Script", variants: [v(400, "pacifico/Pacifico-Regular")] },
  { id: "jetbrains-mono", label: "JetBrains Mono", category: "Mono", variants: [v(400, "jetbrains-mono/JetBrainsMono-Regular"), v(700, "jetbrains-mono/JetBrainsMono-Bold")] },
];

export function findTimelineFont(familyId: string): TimelineFontFamily | undefined {
  return TIMELINE_FONTS.find((f) => f.id === familyId);
}

export function timelineFontCssFamily(familyId: string): string {
  return `TimelineFont_${familyId.replace(/-/g, "_")}`;
}

const loaded = new Set<string>();

/** Lazy-load a font variant for preview via FontFace API. */
export async function ensureTimelineFontLoaded(
  familyId: string,
  weight: number
): Promise<string> {
  const family = findTimelineFont(familyId) ?? TIMELINE_FONTS[0];
  const variant =
    family.variants.find((x) => x.weight === weight) ?? family.variants[0];
  const cssFamily = timelineFontCssFamily(family.id);
  const key = `${cssFamily}-${variant.weight}`;
  if (loaded.has(key)) return cssFamily;

  try {
    const src = variant.woff2;
    const face = new FontFace(cssFamily, `url(${src}), url(${variant.ttf})`, {
      weight: String(variant.weight),
      style: variant.style,
    });
    await face.load();
    document.fonts.add(face);
    loaded.add(key);
  } catch {
    try {
      const face = new FontFace(cssFamily, `url(${variant.ttf})`, {
        weight: String(variant.weight),
        style: variant.style,
      });
      await face.load();
      document.fonts.add(face);
      loaded.add(key);
    } catch {
      /* fallback to sans-serif */
    }
  }
  return cssFamily;
}
